mediaplatform/models.py

```diff
@@ -0,0 +1,146 @@
+
+    # We use on_delete=models.PROTECT here to make sure that the referenced permission cannot be
+    # deleted from under our feet. We set the related name to "+" to stop Django trying to create a
+    # related field back from the Permission object. If it tried to do so there would be a sea of
+    # "allows_view_on_media_item"-like fields on Permission which would clutter things immensely.
+
+    #: :py:class:`~.Permission` determining who can view this item
+    view_permission = models.OneToOneField(Permission, on_delete=models.PROTECT,
+                                           help_text='Restriction on who can see/hear this item',
+                                           related_name='+')
+
+    #: :py:class:`~.Permission` determining who can edit this item
+    edit_permission = models.OneToOneField(Permission, on_delete=models.PROTECT,
+                                           help_text='Restriction on who can modify this item',
+                                           related_name='+')
+
+    # We use max_length=10 here to future proof us against the day when we inevitably have our
+    # first "length 4 three-letter code". :)
+
+    #: ISO 693-3 language code
+    language = models.CharField(
+        max_length=10,
+        help_text=(
+            'ISO 639-3 three letter language code describing majority language used in this item'
+        )
+    )
+
+    #: Video copyright
+    copyright = models.TextField(blank=True, help_text='Free text describing Copyright holder')
+
+    #: List of tags for video
+    tags = pgfields.ArrayField(models.CharField(max_length=256), blank=True,
+                               help_text='Tags/keywords for item')
+
+    #: Creation time
+    created_at = models.DateTimeField(auto_now_add=True)
+
+    #: Last update time
+    updated_at = models.DateTimeField(auto_now=True)
+
+    #: Deletion time. If non-NULL, the item has been "deleted" and should not usually be visible.
+    deleted_at = models.DateTimeField(null=True, blank=True)
+
+
+class Collection(models.Model):
+    """
+    A collection of media items.
+
+    """
+    #: Primary key
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+
+    #: Collection title
+    title = models.TextField(help_text='Title of collection')
+
+    #: Collection description
+    description = models.TextField(help_text='Description for collection')
+
+    #: :py:class:`~.MediaItem` objects which make up this collection. Postgres does not (currently)
+    #: allow array elements to have a foreign key constraint added to them so we need to represent
+    #: the links as bare UUIDs. The upshot of this is that code which uses this field needs to
+    #: handle the (rare) case that a UUID in the list does not correspond to a current video.
+    #: YouTube, as an example, has this problem as well since videos in playlists are sometimes
+    #: replaced by a "deleted video" placeholder.
+    media_items = pgfields.ArrayField(
```

Comment:
why not just a foreingkey in media_item? Is it because we may want to preserve the order? I think is worth adding it to the comments.

Reply:
1) If a foreign key in media_item then a media item can be in one and only one collection which seems a needless restriction.
2) If a foreign key in media_item then there is no ordering

Looking at the way the SMS currently does it, there is a separate table mapping media to collections with an index field. Although this would work, it seems ugly. If you're unhappy with losing the foreign key constraint, I'd use a many-to-many mapping table like this instead.